Lang/CodeGen/NativeTypes/Bool.cs

```diff
@@ -63,6 +67,181 @@ private void BuildAnd(CodeGen.LLVMExposingCodeGen g)
+        var u1 = builder.BuildStructGEP2(Bool, pThis, 0, "unbox1");
+        var u2 = builder.BuildStructGEP2(Bool, pOther, 0, "unbox2");
+        var v1 = builder.BuildLoad2(ctx.Int1Type, u1, "val1");
+        var v2 = builder.BuildLoad2(ctx.Int1Type, u2, "val2");
+        LLVMValueRef res = builder.BuildXor(v1, v2, "xor");
+        /* resptr points both to the structure and to its initial field */
+        LLVMValueRef resptr = builder.BuildMalloc(Bool, "resptr");
+        builder.BuildStore(res, resptr);
+        builder.BuildRet(resptr);
+
+        func.VerifyFunction(LLVMVerifierFailureAction.LLVMPrintMessageAction);
+    }
+
+    private void BuildNot(CodeGen.LLVMExposingCodeGen g)
+    {
+        var ctx = g.ctx;
+        var module = g.module;
+        var builder = g.builder;
+        var FuncType = g.FuncType;
+        var Struct = g.Struct;
+
+        var Bool = Struct["Bool"];
+        var PBool = LLVMTypeRef.CreatePointer(Bool, 0);
+
+        const string funcN = "Bool.Not..Bool";
+        LLVMValueRef func = module.GetNamedFunction(funcN);
+        if (func.BasicBlocks.Length != 0)
+        {
+            throw new CodeGen.LLVMGenException("Function " + funcN + " already has a body");
+        }
+
+        var parames = new LLVMTypeRef[]
+        {
+            /*this*/PBool
+        };
+
+        var funcT = FuncType[funcN] = LLVMTypeRef.CreateFunction(PBool, parames);
+        func = module.AddFunction(funcN, funcT);
+        func.Linkage = LLVMLinkage.LLVMExternalLinkage;
+
+        var pThis = func.GetParam(0);
+        pThis.Name = "this";
+
+        builder.PositionAtEnd(func.AppendBasicBlock("entry"));
+        var u1 = builder.BuildStructGEP2(Bool, pThis, 0, "unbox1");
+        var v1 = builder.BuildLoad2(ctx.Int1Type, u1, "val1");
+        LLVMValueRef res = builder.BuildNot(v1, "not");
+        /* resptr points both to the structure and to its initial field */
+        LLVMValueRef resptr = builder.BuildMalloc(Bool, "resptr");
+        builder.BuildStore(res, resptr);
+        builder.BuildRet(resptr);
+
+        func.VerifyFunction(LLVMVerifierFailureAction.LLVMPrintMessageAction);
+    }
+
+    private void BuildToInteger(CodeGen.LLVMExposingCodeGen g)
+    {
+        var ctx = g.ctx;
+        var module = g.module;
+        var builder = g.builder;
+        var FuncType = g.FuncType;
+        var Struct = g.Struct;
+
+        var Bool = Struct["Bool"];
+        var PBool = LLVMTypeRef.CreatePointer(Bool, 0);
+
+        const string funcN = "Bool.ToInteger..Int";
+        LLVMValueRef func = module.GetNamedFunction(funcN);
+        if (func.BasicBlocks.Length != 0)
+        {
+            throw new CodeGen.LLVMGenException("Function " + funcN + " already has a body");
+        }
+
+        var parames = new LLVMTypeRef[]
+        {
+            /*this*/PBool
+        };
+
+        var funcT = FuncType[funcN] = LLVMTypeRef.CreateFunction(ctx.Int32Type, parames);
```

Comment:
That's not quite correct. `ctx.Int32Type` is a native 32-bit integer type. It can be used for the internal implementation of a class representing numbers (just like `ctx.Int1Type` is used for the internal implementation of `Bool`) but everything exposed to ChineseObjects must work with references to class objects (note the return types of `Bool.And`, `Build.Or`, `Build.Xor`: they all return a pointer to the `Bool` struct rather than a `ctx.Int1Type`.

The same should happen here, except it's only possible after standard types to represent numbers are implemented